### DISCUSSION TITLE
My places improvements

### DISCUSSION THREAD

**Discussion by bios4** [devemux86/cruiser#409] [Q&A] [answered]
1) "My places": if a (e.g. track) group is deleted, it is actually deleted immediately without any security question - and not only after confirmation with the "Save" button!

So could you please either
a) implement a security question if the deletion should really be executed or
b) apply the deletion only after the "Save" button is pressed

My assumption was that all my actions are only really saved/executed when selecting the "Save" button.
Which is somehow meaningless at all in its current state, as all the actions performed are also done without pressing it.

2) "My places > Tracks": importing a file twice should trigger a "File already existing - Overwrite / Cancel?" message instead of importing it without any message and showing it doubled in the list.

**Comment by devemux86** [comment]
@bios4 thanks for the suggestions.

> apply the deletion only after the "Save" button is pressed

1. That's how it's always worked: all changes are applied after pressing the save button
(or back button - implemented after another suggestion).

**Reply by bios4** [reply]
But in my current case the group was deleted IMMEDIATELY, WITHOUT confirming with the "Save"-button!
And without any security question!

Saving when selecting the back button is not optimal too from my POV...
Could lead to unwanted deletions IMHO.

"Back" and "X" should only do what their name says - "go back" and "close" (w/o any deletion, when this is not confirmed - either by a security question or the "Save"-button).

**Reply by devemux86** [reply]
We had security question on each delete, but users did not like it, because it made deletions very slow to perform.

> But in my current case the group was deleted IMMEDIATELY, without confirming with the save button!

@bios4 Are you sure? I cannot confirm this. Can you record a video with your exact steps?

The deletion happens on the list, but it is not saved on storage until save or back button are pressed.
If you press the `X` button and return to the list, it still contains the item.

**Reply by devemux86** [reply]
I can put a question for save in the back arrow button.

**Reply by bios4** [reply]
I'm sure that my items were deleted w/o confirming with the "Save"-button, but can not 100% confirm if I closed the list ("X") or selected the "Back"-button... 😞

> I can put a question for save in the back arrow button.

I would really appreciate this!

**Reply by devemux86** [reply]
@bios4 it will be available in the next Cruiser update.

**Reply by bios4** [reply]
I also encountered a strange issue when exporting my tracks...

I have four track groups - "WOW-Tracks", "2021", "2022" and "2023".
Each of them containing some tracks:

<img alt="" width="33%" src="https://github.com/devemux86/cruiser/assets/145460147/bf812ee3-148a-4f82-93a2-acbfebfe8a62">

When long-pressing the list and selecting "Export all", a zip file is generated but at the same time an error message is shown: "duplicate entry: 2023/Track from GPX.gpx" (see above).

The resulting zip file then contains only two of my four groups - "WOW-Tracks" and "2023".
The groups 2021 and 2022 and the containing tracks are missing in the zip file:

<img alt="" width="33%" src="https://github.com/devemux86/cruiser/assets/145460147/2000dc66-2747-44ff-b049-2a8bab9d540e">

The whole folder looks like this:

<img alt="" width="33%" src="https://github.com/devemux86/cruiser/assets/145460147/609c3e2c-b84c-49a1-9226-a544ac8d716f">

Any ideas @devemux86?

**Reply by devemux86** [reply]
@bios4 can you show images of Cruiser with the track names in each group?

Or what storage folder have you selected in Cruiser settings?
Can you access it with a file manager, zip the "tracks" sub-folder and send it to me?

**Reply by bios4** [reply]
> Or what storage folder have you selected in Cruiser settings?

<img alt="" width="33%" src="https://github.com/devemux86/cruiser/assets/145460147/2e0f7ecf-b4a4-4ef5-a8b7-b9d250b38d39">

> Can you access it with a file manager, zip the "tracks" sub-folder and send it to me?

You've got mail!

**Reply by devemux86** [reply]
> You've got mail!

@bios4 thanks!

So the group "2023" has tracks with the same name "Track from GPX".
As there can't be 2 files with same name in a folder, the zip process complains about it.
We need to improve this by probably adding a suffix to each  file with the same name.

**Reply by devemux86** [reply]
@bios4 when did you import these tracks with same name?
Or what process did you use to add them in favorites?

Latest Cruiser should not allow entering same names.

**Reply by bios4** [reply]
Ah I see...

I added these tracks with a version < 3.0.37 final for sure.

These are Cruiser recordings, which were updated with an external app called "Tracklia".

The files itself were not even "named" with "Tracks from GPX", but e.g. "Rec_J6_20230910_185732_cleaned.gpx".

But all of them contain the `<name>Track from GPX</name>` markup (added automatically by Tracklia unfortunately). This is then taken over by Cruiser, but not the file name...

**Reply by devemux86** [reply]
> This is then taken over by Cruiser, but not the file name...

We can have it one way or the other, which do you prefer?

**Reply by bios4** [reply]
> We can have it one way or the other, which do you prefer?

Hmm, difficult to answer...

But as long as Cruiser's own exported GPS recordings do NOT contain any `<metadata><name>XYZ</name></metadata>` markup (see below - this is what the header of a Cruiser GPS export looks like), I suggest to use the filename as track name.

`<?xml version='1.0' encoding='UTF-8' standalone='no' ?>
<gpx xmlns="http://www.topografix.com/GPX/1/1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.topografix.com/GPX/1/1 http://www.topografix.com/GPX/1/1/gpx.xsd" version="1.1" creator="Cruiser">
  <metadata />
  <trk>
    <trkseg>
    [...]`

What do others think?

**Reply by devemux86** [reply]
@bios4 thanks, we do not set name in the gpx tags for the GPS recording export?

I will add it.
If you see any other cases, please let me know.

**Reply by bios4** [reply]
> [...] we do not set name in the gpx tags for the GPS recording export?

Nope.
What is shown above is the complete header of a Cruiser GPS export file...

**Reply by WolfgangZ42** [reply]
> @devemux86 We can have it one way or the other, which do you prefer?

> @bios4 Hmm, difficult to answer...

For me it's not difficult - please take the filename (e.g. 20231030_175735.gpx) because that's what I can change easily. And the name in the gpx header couldn't be any other - could it?

**Reply by bios4** [reply]
> And the name in the gpx header couldn't be any other - could it?

The name in the header can be whatever the program/app who stores it defines.
Exactly that happened with my files - see my posts above...

And Cruiser for instance (up to now) defines no `<name>`-attribute in the header at all.
Additionally, a single gpx file could basically have multiple tracks in it, and each of these tracks could have an own `<name>`-attribute.

This also needs to be considered somehow.

You see: "for you", it seems not be difficult.
In fact it is a little bit more complex. 😉

**Reply by WolfgangZ42** [reply]
By "couldn't be any other" I meant that Cruiser can't invent a name and we can't specify one - so it can only be a timestamp name. I did not consider what you "manage" with other tools.

Otherwise it already annoys me that BaseCamp (only installed as a test!) takes the name from the header and not the filename. Multiple versions then all keep the original header name. No longer distinguishable in Navigator VI for instance!

Btw: I'm not a track collector at all - so this topic isn't important to me.

**Reply by bios4** [reply]
> By "couldn't be any other" I meant that Cruiser can't invent a name

Sure it could - as the "Tracklia" app was doing with the files I mentioned. They did not have a `name`-attribute, so the app added "Track from GPX", regardless the file's name.

> BaseCamp (only installed as a test!) takes the name from the header and not the filename

It would be interesting to see what Basecamp shows for files without the name in the header...

> I'm not a track collector at all - so this topic isn't important to me.

I'm also no "collector" per se - but I found the "My tracks" feature to be potentially a suitable way of adding my driven routes to a sort of "heatmap". Like Calimoto is offering.

To not overload the map I thought of managing the tracks on a yearly basis, so I can show/hide them at once and determine easily the spots I did not yet go to...

At least this is my long-term goal. 😀

**Reply by devemux86** [reply]
> And Cruiser for instance (up to now) defines no <name>-attribute in the header at all.

This currently seems to happen in GPS recording export.
The other GPX exports should store the name in the tags.

**Reply by devemux86** [reply]
> It would be interesting to see what Basecamp shows for files without the name in the header...

Cruiser checks the tag name and uses the filename if it's empty.
So we just need to skip the check and always use the filename. :slightly_smiling_face:

**Reply by devemux86** [reply]
> I found the "My tracks" feature to be potentially a suitable way of adding my driven routes to a sort of "heatmap". Like Calimoto is offering.
To not overload the map I thought of managing the tracks on a yearly basis, so I can show/hide them at once and determine easily the spots I did not yet go to...

Very nice use of the favorite tracks! :+1:

**Reply by 0709wiwiwi** [reply]
The gpx file name is separate from the internal gpx track name.

The following timestamps are present in a Locus gpx export.
The timestamp (start of record) is used as the track name.
Applying this automatically in this way is quite common.

Example of a track export by Locus.

Export gpx file name: 2023-07-20_12-51.gpx
Internal gpx track name: 2023-07-20 12:51
This notation is in **local time.**

In the track points the timestamp notation is in UTC time.
In the header metadata section you find another UTC time notation.

![image](https://github.com/devemux86/cruiser/assets/54167201/144efe5b-8031-4dfa-aad1-4d1a5abbdce7)
This is the time of the actual Locus gpx track export generation.
If you generate the gpx export again from the Locus track record database, this timestamp will be updated.

Note.
External apps should not simply change **an already EXISTING** track name. Only users can and do, of course. 
If it turns out that external apps unexpectedly change the track name, you could additionally add this in trk cmt ?
Necessary? However, adding additional protection against conditions such as, if if if, is probably quite endless.
![image](https://github.com/devemux86/cruiser/assets/54167201/781a8217-9866-4232-98c4-13231541483a)

**Reply by bios4** [reply]
> This currently seems to happen in GPS recording export.

Yes, this is what I was talking about in this thread.

> If you generate the gpx export again from the Locus track record database, this timestamp will be updated.

From my POV the datetime information in a track/recording file's header should be the CREATION datetime of the track/recording, and not the EXPORT datetime. So I hope that Cruiser would NOT mimic this behaviour (if datetime will be added to the gpx header of the GPS recording export at all)...

**Reply by devemux86** [reply]
However, GPS recording is not a one-time operation.

Many users collect their tracks in one recording that can last several days or weeks.
They don't clean it at every recording start. What is the creation time in such a case?

**Reply by 0709wiwiwi** [reply]
hmm... I do not ask to attach this in a Cruiser gpx track export.
I do show only how Locus attaches this extra information. 
It is a quite harmless extra that sits in that metadata section. 
Users (except the nerds) will hardly  even notice this extra.
Anyway what I prefer is:
Nicely compact & functional.
With a useful gpx track name.
![image](https://github.com/devemux86/cruiser/assets/54167201/9ab009af-9dd7-4842-bc7c-70916b5405fc)

**Reply by bios4** [reply]
> Many users collect their tracks in one recording that can last several days or weeks.

Good point - I wasn't even aware of/didn't think of that possibility... 🤔
But then: leave it from the header at all.

**Reply by 0709wiwiwi** [reply]
> What is the creation time in such a case?

Into the gpx name  ? 
Add the (local time) timestamp of the start of the trackrecord.  
The record will simply continue until you manually end it. 
The gpx standard has this well thought out as follows.
Track Segment - <trkseg> - This is a **logical** grouping of track points. 
For example, if you have a GPS that you turn on and off every day on a multi-day trip, each day will be contained in a separate track segment. Within track segments you'll see track points.
Reference: https://hikingguy.com/how-to-hike/what-is-a-gpx-file/

_There's nothing stopping you from storing and organizing track records in a different way. 
You don't have to have a track record continuously running all the time. 
To export a current gpx trackrecord you first will have to stop this one running._

If you want to present the following well, a lot of development work is needed. 
![image](https://github.com/devemux86/cruiser/assets/54167201/36eb66ef-dfdc-49e0-8795-fd44fee8c261)
But the main task of Cruiser is still and above all to offer optimal navigation?

**Reply by bios4** [reply]
> But the main task of Cruiser is still and above all to offer optimal navigation?

Yes, of course!
Therefore I would not necessarily invest too much effort in all these details ^^ - "just" add a proper name (and header at all in case of the recording export files) and we should be good to go.

Everything above this would be icing on the cake... 😀

**Comment by devemux86** [comment]
All the improvements discussed here are available for testing in:
- https://github.com/devemux86/cruiser/discussions/414

**Reply by bios4** [reply]
Thanks a lot @devemux86!

I am currently not able to record any rides to inspect the GPS record header, but the test of the other discussed issues looked promising! 😀

> "My places > Tracks": importing a file twice should trigger a "File already existing - Overwrite / Cancel?" message instead of importing it without any message and showing it doubled in the list.

Importing the same file twice is now not anymore possible, which is good. But I really would like to see at least a toast message from Cruiser that the file already exists (similar to the toast message when a name of a track is already existing)!

Currently the import dialog just closes and Cruiser does literally "nothing" when trying to import an already existing file again...

**Reply by devemux86** [reply]
@bios4 indeed we need some feedback when something is not imported.

For example, we import a zip with many items, some may have the same name as our own items.
Items are essentially files, they cannot be duplicated and are skipped. Ask each one for rename is complicated.
So we need to inform the user about the ones that could not be imported?

"Item" could be favorite group or route or track.

**Reply by bios4** [reply]
> For example, we import a zip with many items, some may have the same name as our own items.
Items are essentially files, they cannot be duplicated and are skipped. Ask each one for rename is complicated.
So we need to inform the user about the ones that could not be imported?

Idea / proposal:
1) In case all can be imported: let the user know via a toast message.
2) In case ONE item is affected: let the user know which one could not be imported via a toast message.
3) In case MULTIPLE items are affected: store them (as a simple text list) in a log file and let the user know via a toast message that multiple items could not be imported and to look at the log file.

Other possibility: allow on a "per affected item"-basis to cancel / overwrite / rename it during import.
But this might indeed be too complex/annoying!?

**Reply by devemux86** [reply]
The simplest is to inform about the not imported because name already exists.
(it will be available in the next update)

**Reply by devemux86** [reply]
Even better we can do it like Firefox saves the same file by automatically adding a number in the name:
- Favorites
- Favorites 2
- Favorites 3
- ...

This way at least everything is imported and we can clean up possible duplicates later.
As there may be groups or routes / tracks with the same name but different content!

**Reply by devemux86** [reply]
@bios4 you can test the above import improvements in the latest:
- https://github.com/devemux86/cruiser/discussions/414

**Comment by devemux86** [comment, marked as answer]
All discussed improvements are implemented in [Cruiser 3.0.38](https://play.google.com/store/apps/details?id=gr.talent.cruiser).
